### PR TITLE
MGMT-18839: Retry downloading boot artifacts

### PR DESCRIPTION
**Pull request by CrystalChun**
Adds a retry when downloading the boot artifacts from assisted service in case there are connection issues or a server disruption.

/cc @carbonin

### CODE REVIEW

File: src/commands/actions/download_boot_artifacts_cmd.go [hunk 1]
```diff
@@ -116,20 +117,28 @@ func createHTTPClient(caCertPath string) (*http.Client, error) {
...
 }
 
-func download(httpClient *http.Client, filePath, url string) error {
-	res, err := httpClient.Get(url)
-	if err != nil {
-		return fmt.Errorf("failed getting %s: %w", url, err)
+func download(httpClient *http.Client, filePath, url string, retry int) error {
+	var downloadErr error
+	var res *http.Response
+	for attempts := 0; attempts < retry; attempts++ {
```
Comment (carbonin) [inline src/commands/actions/download_boot_artifacts_cmd.go:123]: I think maybe add a delay here so we don't do the 5 requests all right after each other.

Comment (CrystalChun) [reply src/commands/actions/download_boot_artifacts_cmd.go:125]: Good idea, thanks! Modified

File: src/commands/actions/download_boot_artifacts_cmd.go [hunk 2]
```diff
@@ -116,20 +119,29 @@ func createHTTPClient(caCertPath string) (*http.Client, error) {
...
+		res, downloadErr = httpClient.Get(url)
+		if downloadErr == nil && (res.StatusCode >= 200 && res.StatusCode < 300) {
+			break
+		}
+		downloadErr = fmt.Errorf("failed downloading boot artifact from %s, status code received: %d, attempt %d/%d, download error: %w",
+			url, res.StatusCode, attempts, retry, downloadErr)
+		log.Warn(downloadErr.Error())
+		time.Sleep(defaultDownloadTimeout)
```
Comment (paul-maidment) [inline src/commands/actions/download_boot_artifacts_cmd.go:133]: Should we call this `defaultDownloadRetryDelay` ?

Comment (carbonin) [reply src/commands/actions/download_boot_artifacts_cmd.go:133]: Indeed, it's not a timeout.

Comment (CrystalChun) [reply src/commands/actions/download_boot_artifacts_cmd.go]: Good idea, thanks! Changed it

**Review by carbonin — APPROVED** [review]

**Pull request merged**